Plutarch/TryFrom.hs

```diff
@@ -1,368 +1,77 @@
-  PData,
-  PIsData (pfromData),
-  pasByteStr,
-  pasConstr,
-  pasInt,
-  pasList,
-  pasMap,
-  pchooseListBuiltin,
-  pdata,
-  pforgetData,
-  pfstBuiltin,
-  ppairDataBuiltin,
-  psndBuiltin,
- )
+-- Not `OVERLAPPABLE` or `OVERLAPPING` on purpose
+instance {-# OVERLAPS #-} PSubtypeDecl a a
 
-import Plutarch.TermCont (TermCont (TermCont, runTermCont), tcont, unTermCont)
+-- FIXME: Relax subtyping constraint to a @c@ rather than all @c@.
+-- This is currently not really possible because a type must have
+-- exactly one super type and no more.
+instance {-# OVERLAPS #-} (PSubtypeDecl a b, PInner c POpaque ~ b) => PSubtypeDecl a c
 
-import Plutarch.DataRepr.Internal (PIsDataRepr (PIsDataReprRepr), PIsDataReprInstances)
-
-import Plutarch.Reducible (Reducible (Reduce))
+{- | @PSubtype a b@ constitutes a subtyping relation between @a@ and @b@.
+ This concretely means that `\(x :: Term s b) -> punsafeCoerce x :: Term s a`
+ is legal and sound.
 
-import Data.Functor.Const (Const)
+ You can not make an instance for this yourself.
+ You must use the 'PInner' type family of 'PlutusType' to get this instance.
 
-import Data.Kind (Type)
+ Caveat: Only @PInner a POpaque@ is considered unfortunately, as otherwise
+ getting GHC to figure out the relation with multiple supertypes is quite hard.
 
-import Data.Coerce (coerce)
-
------------------------ The class PTryFrom ----------------------------------------------
+ Subtyping is transitive.
+-}
+type PSubtype :: PType -> PType -> Constraint
+type PSubtype = PSubtypeDecl
 
 {- |
-    This checks the data structure for validity.
-    If you don't care about parts of the structure
-    don't verify those parts, just put a `PData` at
-    the places you don't care about.
-    Be aware this might get really expensive, so only
-    use it if you cannot establish trust otherwise
-    (e.g. via only checking a part of your Data with
-    PTryFrom)
-    Laws:
-      - the operation `ptryFrom` mustn't change the representation of the underlying data
-      - the operation `ptryFrom` must always prove the integrity of the whole target type
-        - example:
-          `ptryFrom PData (PAsData (PBuiltinList PData))` must only succeed if the underlying
-          representation is a `BuiltinList` containing any `PData`
-        - all conversions are fallible, this happens if the representation doesn't match
-          the expected type.
-      - the operation `ptryFrom @b @a` proves equality between the less expressive `PType` `a` and
-        the more expressive `PType` `b`, hence the first element of the resulting Tuple
-        must always be wrapped in `PAsData` if the origin type was `PData` (see law 1)
-      - the result type `b` must always be safe than the origin type `a`, i.e. it must carry
-        more information
+@PTryFrom a b@ represents a subtyping relationship between @a@ and @b@,
+and a way to go from @a@ to @b@.
+Laws:
+- @(punsafeCoerce . fst) <$> tcont (pdowncast x) ≡ pure x@
```

Comment:
this states the law wrt pdowncast, is that intended?

Reply:
nope